File: python_development/System/SystemClass.py
import json

from python_development.Advisor import Advisor
from python_development.CourseSession import CourseSession
from python_development.Lecturer import Lecturer
from python_development.Page.LoginPage import LoginPage
from python_development.Page.PageType import PageType
from python_development.Student import Student
from python_development.System.SystemDomain import SystemDomain
from python_development.Page import PageType
import time


class SystemClass:
    def __init__(self, user_interface):
        self.domain = SystemDomain()
        self.currentUser = None
        self.userInterface = user_interface

        login = LoginPage("Welcome! Please enter your username/password.")
        self.userInterface.addPage(login)
        self.userInterface.setCurrentPage("LOGIN_PAGE")

    def run(self):
        while True:
            self.userInterface.display()
            self.listenUserInterface(self.userInterface.getSystemMessage())

    def login(self, user_info):
        all_users = None
        if user_info[0][0] == 'o':
            all_users = self.domain.getStudentCreator().getStudents()
            self.currentUser = Student.login(Student(None, None, None, None,
                                                     None, None, None), user_info, all_users)
        elif user_info[0][0] == 'a':
            all_users = self.domain.getAdvisorCreator().getAdvisors()
            self.currentUser = Advisor.login(Advisor(None, None, None, None),
                                             user_info, all_users)
        elif user_info[0][0] == 'l':
            all_users = self.extract_advisors()
            self.currentUser = Lecturer.login(Lecturer(None, None, None, None),
                                              user_info, all_users)
        if self.currentUser is None:
            print("Username/Password incorrect.\n")
        else:
            self.userInterface.setPages(self.domain.getPageCreator().create_pages(self.currentUser))
            self.userInterface.setCurrentPage("MAIN_MENU_PAGE")
            print("LOGIN SUCCESSFUL - WELCOME {} {}".format(self.currentUser.get_first_name(),
                                                            self.currentUser.get_last_name()))
            all_users.clear()

    def extract_advisors(self):
        lecturer_only = []
        for lecturer in self.domain.getLecturerCreator().getLecturers():
            if not isinstance(lecturer, Advisor):
                lecturer_only.append(lecturer)
        return lecturer_only

    def logout(self):
        self.userInterface.resetPages()
        login = LoginPage("Welcome! Please enter your username/password.")
        self.userInterface.addPage(login)
        self.userInterface.setCurrentPage("LOGIN_PAGE")
        self.setCurrentUser(None)

    def exit(self):
        self.updateStudentJson()
        self.updateCourseJson()
        self.updateAdvisorJson()
        self.updateLecturerJson()
        exit(0)

    def updateCourseJson(self):
        try:
            with open("JSON_Files\\courses.json", 'r') as file:
                content = file.read()
                jsonObject = json.loads(content)
                courseJSON = jsonObject["courses"]
                for i in range(len(courseJSON)):
                    currentCourse = courseJSON[i]
                    currentCourse["quota"] = self.domain.getCourseCreator().getCourses()[i].getQuota()
                    currentCourse["studentList"] = self.studentToJsonArray(
                        self.domain.getCourseCreator().getCourses()[i].getStudentList())
            with open("JSON_Files\\courses.json", 'w') as file:
                file.write(json.dumps(jsonObject, indent=4))
        except (IOError, json.JSONDecodeError):
            print("An error occurred while writing data to the courses JSON file.")

    def updateLecturerJson(self):
        try:
            with open("JSON_Files\\lecturers.json", 'r') as file:
                content = file.read()
                jsonObject = json.loads(content)
                lecturerJSON = jsonObject["lecturers"]
                for i in range(len(self.domain.getLecturerCreator().getLecturers())):
                    if not isinstance(self.domain.getLecturerCreator().getLecturers()[i], Advisor):
                        currentLecturer = lecturerJSON[i]
                        currentLecturer["password"] = self.domain.getLecturerCreator().getLecturers()[
                            i].getPassword().getPassword()
            with open("JSON_Files\\lecturers.json", 'w') as file:
                file.write(json.dumps(jsonObject, indent=4))
        except (IOError, json.JSONDecodeError):
            print("An error occurred while writing data to the lecturers JSON file.")

    def updateAdvisorJson(self):
        try:
            with open("JSON_Files\\advisors.json", 'r') as file:
                content = file.read()
                jsonObject = json.loads(content)
                advisorJSON = jsonObject["advisors"]
                for i in range(len(self.domain.getAdvisorCreator().getAdvisors())):
                    currentAdvisor = advisorJSON[i]
                    currentAdvisor["password"] = self.domain.getAdvisorCreator().getAdvisors()[
                        i].getPassword().getPassword()
            with open("JSON_Files\\advisors.json", 'w') as file:
                file.write(json.dumps(jsonObject, indent=4))
        except (IOError, json.JSONDecodeError):
            print("An error occurred while writing data to the advisors JSON file.")

    def updateStudentJson(self):
        for i in range(len(self.domain.getStudentCreator().getStudents())):
            try:
                studentId = self.domain.getStudentCreator().getStudents()[i].getPersonId().getId()
                with open(f"JSON_Files\\Students\\{studentId}.json", 'r') as file:
                    content = file.read()
                    jsonStudent = json.loads(content)
                    registration = jsonStudent["registration"]
                    selected = self.domain.getStudentCreator().getStudents()[i].getSelectedCourses()
                    approved = self.domain.getStudentCreator().getStudents()[i].getApprovedCourses()
                    registration["selectedCourses"] = self.transcriptCourses(selected)
                    registration["approvedCourses"] = self.transcriptCourses(approved)
                    jsonStudent["password"] = self.domain.getStudentCreator().getStudents()[
                        i].getPassword().getPassword()
                    jsonStudent["request"] = self.domain.getStudentCreator().getStudents()[i].getRequest()
                    jsonStudent["readNotification"] = list(
                        self.domain.getStudentCreator().getStudents()[i].getReadNotifications())
                    jsonStudent["unreadNotification"] = list(
                        self.domain.getStudentCreator().getStudents()[i].getUnreadNotifications())
                with open(f"JSON_Files\\Students\\{studentId}.json", 'w') as file:
                    file.write(json.dumps(jsonStudent, indent=4))
            except (IOError, json.JSONDecodeError):
                print("\nSYSTEM UPDATE FAILS.")

    def transcriptCourses(self, transcript_courses_list):
        transcript_courses_ar = []
        for course in transcript_courses_list:
            if isinstance(course, CourseSession):
                transcript_courses_ar.append(f"{course.getCourseId().getId()}.{course.getSessionId().getId()}")
            else:
                transcript_courses_ar.append(course.getCourseId().getId())
        return transcript_courses_ar

    def studentToJsonArray(self, students):
        student_ids = [student.getPersonId().getId() for student in students]
        return student_ids

    def listenUserInterface(self, sm):
        functionType = sm.getFunctionType()

        if functionType == "LOGIN":
            userInfo = sm.getInput()
            self.login(userInfo)

        elif functionType == "LOGOUT":
            print(
                "LOGOUT SUCCESSFUL - GOODBYE " + self.currentUser.getFirstName() + " " + self.currentUser.getLastName())
            print("bunu renkli yazcan he unutma dayıogli")
            self.logout()
            self.userInterface.setCurrentPage("LOGIN_PAGE")

        elif functionType == "EXIT":
            print("SYSTEM EXITING")
            time.sleep(0.5)
            print("SYSTEM EXITING.")
            time.sleep(0.5)
            print("SYSTEM EXITING..")
            time.sleep(0.5)
            print("SYSTEM EXITING...")
            time.sleep(0.5)
            print("bunu renkli yazcan he unutma dayıogli")
            self.exit()

        elif functionType == "CHANGE_PAGE":
            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "SELECT_COURSE":
            student = self.currentUser
            courseName = student.getSelectableCourses()[int(sm.getInput()) - 1].getCourseName()
            if student.addSelectedCourse(int(sm.getInput())):
                print("Course Addition Is Succesful - " + courseName)
                print("bunu renkli yazcan he unutma dayıogli")
            else:
                print("Course Addition Is Not Succesful - " + courseName)
                print("bunu renkli yazcan he unutma dayıogli")

            selectableCoursePage = self.userInterface.selectPage("SELECTABLE_COURSES_PAGE")
            selectableCoursePage.setContent(
                self.domain.getPageCreator().createSelectableCoursesPageContent(student.getSelectableCourses(),
                                                                                student.getMarks()))
            selectableCoursePage.setNumberOfSelectableCourses(len(student.getSelectableCourses()))

            selectedCoursePage = self.userInterface.selectPage("SELECTED_COURSES_PAGE")
            selectedCoursePage.setContent(
                self.domain.getPageCreator().createSelectedCoursesPageContent(student.getSelectedCourses()))
            selectedCoursePage.setNumberOfDropableCourses(len(student.getSelectedCourses()))

            syllabus = self.userInterface.selectPage("SYLLABUS_PAGE")
            syllabus.setContent(self.domain.getPageCreator().createSyllabusPageContent(student.getSyllabus()))

            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "DROP_COURSE":
            student = self.currentUser
            courseName = student.getSelectableCourses()[int(sm.getInput()) - 1].getCourseName()
            student.dropCourse(int(sm.getInput()))
            print("Course Dropping Is Succesful - " + courseName)
            print("bunu renkli yazcan he unutma dayıogli")

            selectableCoursePage = self.userInterface.selectPage("SELECTABLE_COURSES_PAGE")
            selectableCoursePage.setContent(
                self.domain.getPageCreator().createSelectableCoursesPageContent(student.getSelectableCourses(),
                                                                                student.getMarks()))
            selectableCoursePage.setNumberOfSelectableCourses(len(student.getSelectableCourses()))

            selectedCoursePage = self.userInterface.selectPage("SELECTED_COURSES_PAGE")
            selectedCoursePage.setContent(
                self.domain.getPageCreator().createSelectedCoursesPageContent(student.getSelectedCourses()))
            selectedCoursePage.setNumberOfDropableCourses(len(student.getSelectedCourses()))

            syllabus = self.userInterface.selectPage("SYLLABUS_PAGE")
            syllabus.setContent(self.domain.getPageCreator().createSyllabusPageContent(student.getSyllabus()))

            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "SEND_APPROVE":
            student = self.currentUser
            if student.getRequest() == "false":
                student.sendToApproval()
                print("You have successfully sent your course selection list for advisor approval!")
                print("bunu renkli yazcan he unutma dayıogli")
            else:
                print("You have already successfully sent your course selection list for advisor approval!")
                print("bunu renkli yazcan he unutma dayıogli")
                print("bunu renkli yazcan he unutma dayıogli")

            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "SELECET_STUDENT":
            advisor = self.currentUser
            advisor.selectStudent(int(sm.getInput()))

            selectedStudentRequestPage = self.userInterface.selectPage("SELECTED_STUDENT_REQUEST_PAGE")
            selectedStudentRequestPage.setContent(
                self.domain.getPageCreator().createSelectedStudentsRequestPageContent(advisor.getSelectStudent()))
            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "APPROVE_REQUEST":
            advisor = self.currentUser
            selectedStudentFullName = (advisor.getSelectStudent().getFirstName() + " " +
                                       advisor.getSelectStudent().getLastName())
            advisor.sendNotification(sm.getInput(), "A")
            advisor.approve()

            print("Request Has Been Approved - " + selectedStudentFullName + "'s Request")
            print("bunu renkli yazcan he unutma dayıogli")

            selectedStudentRequestPage = self.userInterface.selectPage("SELECTED_STUDENT_REQUEST_PAGE")
            selectedStudentRequestPage.setContent(
                self.domain.getPageCreator().createSelectedStudentsRequestPageContent(advisor.getSelectStudent()))

            evaluateRequestPage = self.userInterface.selectPage("EVALUATE_REQUESTS_PAGE")
            evaluateRequestPage.setContent(
                self.domain.getPageCreator().createEvaluateRequestPageContent(advisor.getAwaitingStudents()))
            evaluateRequestPage.setNumberOfRequest(len(advisor.getAwaitingStudents()))
            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "DISAPPREOVE_REQUEST":
            advisor = self.currentUser
            selectedStudentFullName = (advisor.getSelectStudent().getFirstName() + " " +
                                       advisor.getSelectStudent().getLastName())
            advisor.sendNotification(sm.getInput(), "R")
            advisor.disapprove()

            print("Request Has Been Disapproved - " + selectedStudentFullName + "'s Request")
            print("bunu renkli yazcan he unutma dayıogli")
            selectedStudentRequestPage = self.userInterface.selectPage("SELECTED_STUDENT_REQUEST_PAGE")
            selectedStudentRequestPage.setContent(
                self.domain.getPageCreator().createSelectedStudentsRequestPageContent(advisor.getSelectStudent()))

            evaluateRequestPage = self.userInterface.selectPage("EVALUATE_REQUESTS_PAGE")
            evaluateRequestPage.setContent(
                self.domain.getPageCreator().createEvaluateRequestPageContent(advisor.getAwaitingStudents()))
            evaluateRequestPage.setNumberOfRequest(len(advisor.getAwaitingStudents()))
            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "SELECT_MY_COURSE":
            lecturer = self.currentUser
            lecturer.selectCourse(int(sm.getInput()))

            selectedMyCoursePage = self.userInterface.selectPage("SELECTED_MY_COURSE_PAGE")
            selectedMyCoursePage.setContent(
                self.domain.getPageCreator().createSelectedMyCoursePage(lecturer.getSelectedCourse()))

            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "CHANGE_PASSWORD":
            passwords = sm.getInput()

            if self.currentUser.getPassword().compareCurrentPassword(passwords[0]):
                if self.currentUser.getPassword().checkPasswordCond(passwords[1]):
                    self.currentUser.getPassword().setPassword(passwords[1])
                    print("Password Change Successful")
                    print("bunu renkli yazcan he unutma dayıogli")
                else:
                    print("Your new password must obey the rules!")
                    print("bunu renkli yazcan he unutma dayıogli")
            else:
                print("Your current password incorrect!")
                print("bunu renkli yazcan he unutma dayıogli")

            self.userInterface.setCurrentPage(sm.getNextPageType())

        elif functionType == "READ_NOTIFICATIONS":
            student = self.currentUser
            student.clearUnreadNotification()

            mainMenuPageStudent = self.userInterface.selectPage("MAIN_MENU_PAGE")
            mainMenuPageStudent.setContent(
                self.domain.getPageCreator().createMainMenuPageStudentContent(len(student.getUnreadNotifications())))

            self.userInterface.setCurrentPage(sm.getNextPageType())

    def getCurrentUser(self):
        return self.currentUser

    def setCurrentUser(self, currentUser):
        self.currentUser = currentUser

File: iter1/src/SystemClass.java
// Press Shift twice to open the Search Everywhere dialog and type `show whitespaces`,
// then press Enter. You can now see whitespace characters in your code.

import systemMessagePackage.*;
import userInfoPackage.*;
import userInterfacePackage.*;
import pagePackage.*;


import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;


import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;

public class SystemClass {
    private SystemDomain domain;
    private Person currentUser;
    private PageContentCreator pageContentCreator;
    private UserInterface userInterface;

    //Constructor
    public SystemClass(UserInterface u) throws JSONException, IOException {
        pageContentCreator = new PageContentCreator();
    	userInterface = u;
		LoginPage login = new LoginPage("welcome");
		userInterface.addPage(login);
		userInterface.setCurrentPage(PageType.LOGIN_PAGE);
    	domain = new SystemDomain();
    }

    //Start running the code
	public void run() {
		while (true) {
			userInterface.display();
			listenUserInterface(userInterface.getSystemMessage());
		}
	}

    //Login with the given user info
    public void login(UserInfo userInfo) {
        boolean userFound = false;
        if (userInfo.getUsername().charAt(0) == 'o') {
            ArrayList<Student> students = domain.getStudents();
            for (Student student : students) {
                if (("o" + student.getStudentId().getId()).equals(userInfo.getUsername()) &&
                        student.getPassword().getPassword().equals(userInfo.getPassword())) {
                    setCurrentUser(student);
                    userFound = true;
                    userInterface.setPages(domain.createPages(currentUser));
                    userInterface.setCurrentPage(PageType.MAIN_MENU_PAGE_STUDENT);
                    break;
                }
            }
        } else if (userInfo.getUsername().charAt(0) == 'a') {
            ArrayList<Advisor> advisors = domain.getAdvisors();
            for (Advisor advisor : advisors) {
                if (("a" + advisor.getLecturerId().getId()).equals(userInfo.getUsername()) &&
                        advisor.getPassword().getPassword().equals(userInfo.getPassword())) {
                    advisor.findAwaitingStudents();
                    System.err.println(advisor.getAwaitingStudents().size());
                	setCurrentUser(advisor);
                    userFound = true;
                    userInterface.setPages(domain.createPages(currentUser));
                    userInterface.setCurrentPage(PageType.MAIN_MENU_PAGE_ADVISOR);
                    break;
                }
            }
        } if (!userFound) {
            System.out.println("Username/Password incorrect.");
        }
    }

    //Logout from an account
    public void logout() {
    	userInterface.resetPages();
    	LoginPage login = new LoginPage("welcome");
		userInterface.addPage(login);
		userInterface.setCurrentPage(PageType.LOGIN_PAGE);
        setCurrentUser(null);
    }

    //Exit the system
    public void exit() throws JSONException, IOException {
        updateStudentJSON();
        updateCourseStudentData();
        updateCourseQuotaData();
        System.exit(0);
    }

    //Update course quota after selections in JSON files
    public void updateCourseQuotaData() throws IOException, JSONException {
        String content = null;
        content = new String(Files.readAllBytes(Path.of("src\\JSON_Files\\courses.json")));
        JSONObject jsonObject = new JSONObject(content);
        JSONArray courseJSON = jsonObject.getJSONArray("courses");

        int j=0;
        for(int i=0; i<courseJSON.length();i++){
            JSONObject curr = courseJSON.getJSONObject(i);
            if(curr.getBoolean("hasSession")){
                JSONArray currCourseJSON = courseJSON.getJSONObject(i).getJSONArray("session");
                int k=0;
                for(k=0;k<currCourseJSON.length();k++){
                    JSONObject currSession = currCourseJSON.getJSONObject(k);
                    currSession.put("quota",domain.getCourses().get(j).getQuota());
                }
                j=j+k;
            }
            else{
                curr.put("quota",domain.getCourses().get(j).getQuota());
                j++;
            }
        }
        Files.write(Paths.get("src\\JSON_Files\\courses.json"),jsonObject.toString(4).getBytes(),StandardOpenOption.TRUNCATE_EXISTING);

    }

    //Update courses student lists after selection in JSON files
    public void updateCourseStudentData() throws IOException, JSONException {
        String content = null;
        content = new String(Files.readAllBytes(Path.of("src\\JSON_Files\\courses.json")));
        JSONObject jsonObject = new JSONObject(content);
        JSONArray courseJSON = jsonObject.getJSONArray("courses");

        int j=0;
        for(int i=0; i<courseJSON.length();i++){
            JSONObject curr = courseJSON.getJSONObject(i);
            if(curr.getBoolean("hasSession")){
                JSONArray currCourseJSON = courseJSON.getJSONObject(i).getJSONArray("session");
                int k=0;
                for(k=0;k<currCourseJSON.length();k++){
                    JSONObject currSession = currCourseJSON.getJSONObject(k);
                    currSession.put("studentList",studentToJsonArray(domain.getCourses().get(j).getStudentList()));
                }
                j=j+k;
            }
            else{
                curr.put("studentList",studentToJsonArray(domain.getCourses().get(i).getStudentList()));
                j++;
            }
        }
        Files.write(Paths.get("src\\JSON_Files\\courses.json"),jsonObject.toString(4).getBytes(),StandardOpenOption.TRUNCATE_EXISTING);
    }

    //Update student infos in JSON files
    public void updateStudentJSON() throws JSONException, IOException {
        for (int i = 0; i < domain.getStudents().size(); i++) {
            String studentId = domain.getStudents().get(i).getStudentId().getId();
            Path path = Path.of("src\\JSON_Files\\" + studentId + ".json");
            String content = new String(Files.readAllBytes(path));
            JSONObject jsonStudent = new JSONObject(content);
            JSONObject registration = jsonStudent.getJSONObject("registration");

            ArrayList<Course> selected = domain.getStudents().get(i).getSelectedCourses();
            ArrayList<Course> approved = domain.getStudents().get(i).getApprovedCourses();

            registration.put("selectedcourses",transcriptCourses(selected));
            registration.put("approvedcourses", transcriptCourses(approved));

            jsonStudent.put("request", domain.getStudents().get(i).getRequest());
            jsonStudent.put("notification", domain.getStudents().get(i).getNotification());
            Files.write(path, jsonStudent.toString(4).getBytes(), StandardOpenOption.TRUNCATE_EXISTING);
        }
    }

    //Create an array for JSON files update
    private String[] studentToJsonArray(ArrayList<Student> students){
        String[] studentIds = new String[students.size()];
        for(int i=0; i<students.size();i++){
            studentIds[i] = students.get(i).getStudentId().getId();
        }
        return  studentIds;
    }

    //Create an array for JSON files update
    private String[] transcriptCourses(ArrayList<Course> transcriptCoursesList) {
        String[] transcriptCoursesAr = new String[transcriptCoursesList.size()];
        for (int i = 0; i < transcriptCoursesList.size(); i++) {
            if (transcriptCoursesList.get(i) instanceof CourseSession) {
                CourseSession crsSession = (CourseSession) transcriptCoursesList.get(i);
                transcriptCoursesAr[i] = crsSession.getCourseId().getId() + "." + crsSession.getSessionId().getId();
            } else {
                transcriptCoursesAr[i] = transcriptCoursesList.get(i).getCourseId().getId();
            }
        } return transcriptCoursesAr;
    }

    //PAGE MERGE WITH SYSTEM PROCESSES
    public void listenUserInterface(SystemMessage sm) {
        FunctionType functionType = sm.getFunctionType();

        if (functionType == FunctionType.NONE) {
            return;
        }

        if (functionType == FunctionType.LOGIN) {
            UserInfo userInfo = (UserInfo)sm.getInput();
            this.login(userInfo);
            return;
        }

        if (functionType == FunctionType.LOGOUT) {
			this.logout();
            this.userInterface.setCurrentPage(PageType.LOGIN_PAGE);
        }


        if (functionType == FunctionType.EXIT) {
            try {
            	this.exit();
            }
        	catch (Exception e){
        	}
        }

        if (functionType == FunctionType.CHANGE_PAGE) {
            this.userInterface.setCurrentPage(sm.getNextPageType());
            return;

        }

        if (functionType == FunctionType.SELECT_COURSE ) {

			Student student = (Student) this.getCurrentUser();
			student.addSelectedCourse((Integer)sm.getInput());

            // handling of selecteable course data
            SelectableCoursesPage selectableCoursePage = (SelectableCoursesPage) this.userInterface.selectPage(PageType.SELECTABLE_COURSES_PAGE);
			selectableCoursePage.setContent(this.pageContentCreator.createSelectableCoursesPageContent(student.getSelectableCourses(), student.getSelectedCourses()));
			selectableCoursePage.setNumberOfSelectableCourses(student.getSelectableCourses().size());

            // handling selected course data
            SelectedCoursesPage selectedCoursePage = (SelectedCoursesPage) this.userInterface.selectPage(PageType.SELECTED_COURSES_PAGE);
			selectedCoursePage.setContent(this.pageContentCreator.createSelectedCoursesPageContent(student.getSelectedCourses()));
			selectedCoursePage.setNumberOfDropableCourses(student.getSelectedCourses().size());

            this.userInterface.setCurrentPage(sm.getNextPageType());

            return;

        }

        if (functionType == FunctionType.DROP_COURSE ) {
            Student student = (Student) this.getCurrentUser();

            student.dropCourse((Integer)sm.getInput());

            // handling selected course data
            SelectedCoursesPage selectedCoursePage = (SelectedCoursesPage) this.userInterface.selectPage(PageType.SELECTED_COURSES_PAGE);
			selectedCoursePage.setContent(this.pageContentCreator.createSelectedCoursesPageContent(student.getSelectedCourses()));
			selectedCoursePage.setNumberOfDropableCourses(student.getSelectedCourses().size());


            // handling of selecteable course data
            SelectableCoursesPage selectableCoursePage = (SelectableCoursesPage) this.userInterface.selectPage(PageType.SELECTABLE_COURSES_PAGE);
			selectableCoursePage.setContent(this.pageContentCreator.createSelectableCoursesPageContent(student.getSelectableCourses(), student.getSelectedCourses()));
			selectableCoursePage.setNumberOfSelectableCourses(student.getSelectableCourses().size());

            this.userInterface.setCurrentPage(sm.getNextPageType());

            return;

        }

        if (functionType == FunctionType.SEND_APPROVE ) {
            Student student = (Student) this.getCurrentUser();
            if (student.getRequest().equals("false")) {
                student.sendToApproval();
            }
            this.userInterface.setCurrentPage(sm.getNextPageType());
        }


        if (functionType == FunctionType.SELECET_STUDENT ) {
			Advisor advisor = (Advisor)this.getCurrentUser();
			advisor.selectStudent((Integer)sm.getInput());
			SelectedStudentRequestPage s = (SelectedStudentRequestPage) this.userInterface.selectPage(PageType.SELECTED_STUDENT_REQUEST_PAGE);
            s.setContent(pageContentCreator.createSelectedStudentsRequesPageContent(advisor.getSelectStudent()));
			this.userInterface.setCurrentPage(sm.getNextPageType());
            return;
        }

        if (functionType == FunctionType.APPROVE_REQUEST ) {
			Advisor advisor = (Advisor)this.getCurrentUser();

			advisor.Approve();

            // handling selected student request
            SelectedStudentRequestPage selectedStdudentRequesPage = (SelectedStudentRequestPage) this.userInterface.selectPage(PageType.SELECTED_STUDENT_REQUEST_PAGE);
			selectedStdudentRequesPage.setContent(this.pageContentCreator.createSelectedStudentsRequesPageContent(advisor.getSelectStudent()));

            // handling evaluate request
            EvaluateRequestsPage evaluateRequestPage = (EvaluateRequestsPage) this.userInterface.selectPage(PageType.EVALUATE_REQUESTS_PAGE);
			evaluateRequestPage.setContent(this.pageContentCreator.createEvaluateRequestPageContent(advisor.getAwaitingStudents()));
			evaluateRequestPage.setNumberOfRequest(advisor.getAwaitingStudents().size());

            this.userInterface.setCurrentPage(sm.getNextPageType());

            return;

        }

        if (functionType == FunctionType.DISAPPREOVE_REQUEST ) {

			Advisor advisor = (Advisor)this.getCurrentUser();

			advisor.Disapprove();

            // handling selected student request
            SelectedStudentRequestPage selectedStdudentRequesPage = (SelectedStudentRequestPage) this.userInterface.selectPage(PageType.SELECTED_STUDENT_REQUEST_PAGE);
			selectedStdudentRequesPage.setContent(this.pageContentCreator.createSelectedStudentsRequesPageContent(advisor.getSelectStudent()));

            // handling evaluate request
            EvaluateRequestsPage evaluateRequestPage = (EvaluateRequestsPage) this.userInterface.selectPage(PageType.EVALUATE_REQUESTS_PAGE);
			evaluateRequestPage.setContent(this.pageContentCreator.createEvaluateRequestPageContent(advisor.getAwaitingStudents()));
			evaluateRequestPage.setNumberOfRequest(advisor.getAwaitingStudents().size());

            this.userInterface.setCurrentPage(sm.getNextPageType());
           
            return;
        }
    }

    public SystemDomain getDomain() {
        return domain;
    }

    public Person getCurrentUser() {
        return currentUser;
    }

    public void setCurrentUser(Person currentUser) {
        this.currentUser = currentUser;
    }
}

